core/src/serialize/yaml.rs

```diff
@@ -232,24 +231,95 @@ impl YamlLoader {
-                    )
-                })?,
-            ),
-            Scalar::String(s) => Term::Str(s.into_owned().into()),
+        // Returns Ok(Some(f)) if it succeeds, Ok(None) if the argument wasn't a float, or
+        // Err it it was a float that Nickel doesn't support (i.e. inf or nan).
+        //
+        // The motivation here is that we want an error on ".inf" instead of silently treating
+        // it as a string.
+        let parse_float = |v: &str| -> Result<Option<Term>, ParseError> {
+            match v {
+                ".inf" | ".Inf" | ".INF" | "+.inf" | "+.Inf" | "+.INF" | "-.inf" | "-.Inf"
+                | "-.INF" | ".nan" | ".NaN" | ".NAN" => Err(ParseError::ExternalFormatError(
+                    "yaml".into(),
+                    "Nickel numbers cannot be inf or NaN".into(),
+                    pos.into_opt(),
+                )),
+                // `f64::from_str` will succeed on "NaN" and "inf". Since those aren't
+                // valid YAML floats, we want to skip them.
+                _ if v.as_bytes().iter().any(u8::is_ascii_digit) => {
+                    let f = v.parse::<f64>().ok();
+                    // unwrap: we've already checked for inf and nan
+                    Ok(f.map(|f| Term::Num(Number::try_from_float_simplest(f).unwrap())))
+                }
+                _ => Ok(None),
+            }
+        };
+
+        // Parse a YAML scalar, inferring the type from the value itself.
+        let parse = |v: Cow<'_, str>| -> Result<Term, ParseError> {
+            if let Some(number) = v.strip_prefix("0x") {
+                if let Ok(i) = i64::from_str_radix(number, 16) {
+                    return Ok(Term::Num(i.into()));
+                }
+            } else if let Some(number) = v.strip_prefix("0o") {
+                if let Ok(i) = i64::from_str_radix(number, 8) {
+                    return Ok(Term::Num(i.into()));
+                }
+            } else if let Some(number) = v.strip_prefix('+') {
+                if let Ok(i) = number.parse::<i64>() {
+                    return Ok(Term::Num(i.into()));
+                }
+            }
+            match &*v {
+                "~" | "null" | "Null" | "NULL" => Ok(Term::Null),
+                "true" | "True" | "TRUE" => Ok(Term::Bool(true)),
+                "false" | "False" | "FALSE" => Ok(Term::Bool(false)),
+                _ => {
+                    if let Ok(i) = v.parse::<i64>() {
+                        Ok(Term::Num(i.into()))
+                    } else if let Some(f) = parse_float(&v)? {
+                        Ok(f)
+                    } else {
+                        Ok(Term::Str(v.into()))
+                    }
+                }
+            }
         };
+
+        let t = if style != ScalarStyle::Plain {
+            // Any quoted scalar is a string.
+            Term::Str(value.into())
+        } else if let Some(tag) = tag.map(Cow::as_ref) {
+            if tag.is_yaml_core_schema() {
```

Comment:
By the way, if we switch to edition 2024 (which I did anyway on the RFC007 PR for that very reason), we have let-chains stabilized and I _believe_ we can rewrite this to avoid the duplicated `else { parse(value)? }` and the double nesting:

```suggestion
        } else if let Some(tag) = tag.map(Cow::as_ref) && tag.is_yaml_core_schema() {
```

Reply:
I tried switching to edition 2024, but it caused a bunch of errors because of the [match ergonomics changes](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/match-ergonomics.html). I guess you've already fixed these in your PR, so in order to avoid merge conflicts I'll just mark this as TODO for now.